### ISSUE TITLE
Cannot follow #include_next directives

### ISSUE THREAD

**Issue by ckandeler** [clangd/clangd#1191]
If I have this in a header file myheader.h:
`#include_next <myheader.h>`
clangd will complain "Main file cannot be included recursively when building a preamble".
The construct seems to be handled fine otherwise, e.g. symbols from the "next" header are available for completion in the file that includes the "first" header. It's only in the header itself that clangd does not seem to be understand that we are not including the same file again.

**Comment by HighCommander4** [comment]
Indeed, opening a file containing an `#include_next` directive is not well supported.

The reason for this is that evaluating an `#include_next` directive requires a piece of state ("which entry in the include search path was used when including the current file") which is absent when processing the main file (i.e. the file that is being opened directly), since the main file is not being processed as a result of an inclusion from another file.

This can be seen independently of clangd in an example like the following:

Directory structure:

```
test.cpp
include1/
    test.hpp
include2/
    test.hpp
```

include2/test.hpp:

```
struct A {};
```

include1/test.hpp:

```
#include_next "test.hpp"
struct B {};
```

test.cpp:

```
#include "test.hpp"
```

Now, if you run `clang -Iinclude1 -Iinclude2 test.cpp`, that works fine.

But if you try run run `clang -Iinclude1 -Iinclude2 include1/test.hpp` (which is approximately what clangd does when opening `include1/test.hpp`), clang is not happy and basically appears to go into an infinite loop including `include1/test.hpp` over and over again until it runs into `-ferror-limit`.

My sense is this is not likely to be an easy fix; it would likely involve some amount of rearchitecting of how `#include_next` is handled by clang, to support it in the main file for the sake of clangd's use case.

**Comment by isgursoy** [comment]
This is also the case for the system headers and bringing tons of clangd code model errors everywhere.

- /usr/lib/llvm-16/include/c++/v1/stdint.h
`stdint.h:124:17: Main file cannot be included recursively when building a preamble`

You know what happens then...

**Comment by HighCommander4** [comment]
> this issue is affecting me severely and makes almost any use of clangd for my proejcts really annoying and often impossible as most functionality will not work because it causes clangd to stop emitting errors and thus not having me see the errors and warnings I am really interested in. Headers like stdint.h are affected and used basically everywhere... This also causes it to not know about a lot of types that would be in the other stdint.h that would be included if it worked corerectly.

To clarify, this issue only affects scenarios where the **open file** directly contains an `#include_next` directive. So, e.g. if you open `<stdint.h>` itself, you might see false positive diagnostics while viewing that file.

However, files that include `<stdint.h>` should not be affected and should not have any false positive diagnostics. If you're seeing that, then you're experiencing an unrelated issue; in that case, please file a new issue and share [clangd logs](https://clangd.llvm.org/troubleshooting.html#gathering-logs) for further diagnosis.

**Comment by PlasmaHH** [comment]
Yeah you might be right, I got myself confused with two other bugs happening at once, one where clangd crashes the other where the driver query assembles a non-existent -isystem flag and depending on where I jump I get different errors...
